### PR TITLE
NOX-534 Add validation for entity chain

### PR DESCRIPTION
**Pull request by anna-naumova25**
Will add tests to cover different flows in the next PR. All the existing tests for redirection pass fine.

### CODE REVIEW

File: samples/Cryptocash.Integration/Generated/Nox.Generator/Nox.Generator.NoxCodeGenerator/Application.Queries/ValidateEntityChainQuery.g.cs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+                return false;
+
+            if (!_entityContextPerEntityName.TryGetValue(relatedPluralName, out var relatedContext))
+                return false;
+            
+            query = query.Where($"{relatedContext.KeyName} == {property.NavigationKey}");
+            previousAggregateRoot = relatedPluralName;
+        }
+
+        return query.Count() > 0;
```
Comment (rochar) [inline samples/Cryptocash.Integration/Generated/Nox.Generator/Nox.Generator.NoxCodeGenerator/Application.Queries/ValidateEntityChainQuery.g.cs]: Any so it does not need to get all rows, just one

File: src/Nox.Lib/Application/Queries/IValidateEntityChainQuery.cs [hunk 2]
```diff
@@ -0,0 +1,8 @@
+﻿namespace Nox.Application.Queries;
+
+public interface IValidateEntityChainQueryHandler
```
Comment (rochar) [inline src/Nox.Lib/Application/Queries/IValidateEntityChainQuery.cs]: I think QueryHandler mislead people... this is actually a service

This is  Relationship Validator, maybe IRelationshipChainValidator for its name?

And move it to /Application/Services/...

File: src/Nox.Lib/Application/Queries/IValidateEntityChainQuery.cs [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+{
+    public bool Handle(ValidateEntityChainQuery request);
```
Comment (rochar) [inline src/Nox.Lib/Application/Queries/IValidateEntityChainQuery.cs]:
```suggestion
    public bool IsValid(ValidateEntityChainQuery relationshipChain);
```

File: src/Nox.Lib/Application/Queries/IValidateEntityChainQuery.cs [hunk 4]
```diff
@@ -0,0 +1,8 @@
...
+}
+
+public partial record ValidateEntityChainQuery(string EntityName, string EntityKey, params (string NavigationName, string NavigationKey)[] NavigationProperties);
```
Comment (rochar) [inline src/Nox.Lib/Application/Queries/IValidateEntityChainQuery.cs]:
```suggestion
public partial record RelationshipChain(string EntityName, string EntityKey, params (string NavigationName, string NavigationKey)[] SortedNavigationProperties);
```

File: src/Nox.Lib/Middlewares/RelatedEntityRoutingMiddleware.cs [hunk 5]
```diff
@@ -21,13 +23,16 @@ internal class RelatedEntityRoutingMiddleware
...
     private readonly int _apiPrefixSliceIndex;
     private readonly int _endpointsMaxDepth;
+    private readonly IValidateEntityChainQueryHandler _validator;
     private readonly HashSet<string> _entitiesPluralNamesLowerCase;
     private readonly Dictionary<string, string> _navigationNameToEntityPluralName = new(StringComparer.OrdinalIgnoreCase);
     private readonly IReadOnlySet<(string entityName, string navigationName)> _canRedirect;
     private readonly ArrayPool<string> poolOfStrings = ArrayPool<string>.Shared;
 
-    public RelatedEntityRoutingMiddleware(RequestDelegate next, NoxSolution solution)
+    public RelatedEntityRoutingMiddleware(RequestDelegate next, NoxSolution solution, IValidateEntityChainQueryHandler validator)
```
Comment (rochar) [inline src/Nox.Lib/Middlewares/RelatedEntityRoutingMiddleware.cs]:
```suggestion
    public RelatedEntityRoutingMiddleware(RequestDelegate next, NoxSolution solution, IValidateEntityChainQueryHandler relationshipChainValidator)
```

File: src/Nox.Lib/Middlewares/RelatedEntityRoutingMiddleware.cs [hunk 6]
```diff
@@ -177,6 +185,23 @@ private bool IsFirstPairValid(string entityName, string navigationName)
...
     }
 
+    private bool IsChainValid(IReadOnlyList<string> segments, int segmentCount)
+    {
+        var count = segmentCount;
+        if (segments[count - 1].Equals(_refSegment))
+            count--;
+
+        var navigationPropertiesCount = (count - 3) / 2;
+        var navigationProperties = new (string, string)[navigationPropertiesCount];
```
Comment (rochar) [inline src/Nox.Lib/Middlewares/RelatedEntityRoutingMiddleware.cs]: add name to the parameters tuple for better readability

File: src/Nox.Lib/Middlewares/RelatedEntityRoutingMiddleware.cs [hunk 7]
```diff
@@ -177,6 +185,23 @@ private bool IsFirstPairValid(string entityName, string navigationName)
         return _canRedirect.Contains((entityName, navigationName));
     }
 
+    private bool IsChainValid(IReadOnlyList<string> segments, int segmentCount)
+    {
+        var count = segmentCount;
+        if (segments[count - 1].Equals(_refSegment))
+            count--;
+
+        var navigationPropertiesCount = (count - 3) / 2;
```
Comment (rochar) [inline src/Nox.Lib/Middlewares/RelatedEntityRoutingMiddleware.cs]: please add a comment to explain this.. in a few weeks we will not understand what we were doing :)

File: tests/Nox.ClientApi.Tests/Generated/Nox.Generator/Nox.Generator.NoxCodeGenerator/Application.Queries/ValidateEntityChainQuery.g.cs
Comment (rochar) [inline tests/Nox.ClientApi.Tests/Generated/Nox.Generator/Nox.Generator.NoxCodeGenerator/Application.Queries/ValidateEntityChainQuery.g.cs:1]: @anna-naumova25  we should add some tests for this... no need to call the api but test the Validator directly

**Review by rochar — CHANGES_REQUESTED** [review]
lo=oks good, minor suggestions

File: tests/Nox.ClientApi.Tests/Generated/Nox.Generator/Nox.Generator.NoxCodeGenerator/Application.Queries/ValidateEntityChainQuery.g.cs
Comment (anna-naumova25) [reply tests/Nox.ClientApi.Tests/Generated/Nox.Generator/Nox.Generator.NoxCodeGenerator/Application.Queries/ValidateEntityChainQuery.g.cs:1]: will add in the next PR

**Comment by anna-naumova25** [comment]
@rochar the PR is now updated according to all the comments provided

File: src/Nox.Lib/Configuration/NoxUseOptions.cs [hunk 8]
```diff
@@ -101,9 +101,12 @@ public void Configure(IApplicationBuilder builder)
         builder.UseWhen(context => context.Request.Path.StartsWithSegments(apiPrefix) && solution.Domain is not null,
             appBuilder =>
             {
-                appBuilder
-                    .UseMiddleware<RelatedEntityRoutingMiddleware>()
-                    .UseMiddleware<ApiRoutingMiddleware>();                    
+                if (solution.Presentation.ApiConfiguration.ApiGenerateRelatedEndpointsMaxDepth > 1 &&
```
Comment (rochar) [inline src/Nox.Lib/Configuration/NoxUseOptions.cs:104]: in the next PR we can add this logic to a static function inside RelatedEntityRoutingMiddleware, `RelatedEntityRoutingMiddleware.IsApplicable(solution)`

**Review by rochar — APPROVED** [review]

**Pull request merged**